core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1800,16 +1802,16 @@ class Log(@volatile var dir: File,
       // if we crash in the middle of this we complete the swap in loadSegments()
       if (!isRecoveredSwapFile)
         sortedNewSegments.reverse.foreach(_.changeFileSuffixes(Log.CleanedFileSuffix, Log.SwapFileSuffix))
-      sortedNewSegments.reverse.foreach(addSegment(_))
 
-      // delete the old files
+      // schedule old segment for async deletion and remove them from the segment index
       for (seg <- sortedOldSegments) {
-        // remove the index entry
-        if (seg.baseOffset != sortedNewSegments.head.baseOffset)
-          segments.remove(seg.baseOffset)
-        // delete segment
-        asyncDeleteSegment(seg)
+        if (segments.remove(seg.baseOffset) != null)
```

Comment:
I wonder if this new ordering has any implication on readers, given that they are lockless. Let's say we are compacting two segments with offsets [0, 100] and [101, 200] into a new segment with offset [0, 200]. With this change, it seems like there is a point in time where we remove the entire range of offsets from 0-200 from `segments` which might cause readers from missing this offset range. Do you think this is a valid scenario?

Reply:
In other words, using the same scenario as example, the existing code in Kafka will first add the new segment [0, 200]. At this moment Log.logSegments() will return segments [0, 200] and [101, 200], which contains duplicate messages and this also could be an issue.

Reply:
Hmm I need to look through the code to figure out what the expectation is for callers of `Log.logSegments()`. You are right though, if we care about a consistent view of `segments`, we should grab a lock. I do think not all operations need to be protected, specially the ones in `Log.read()` could continue to be lock-free. May be you could file a separate JIRA to track this investigation?

Reply:
We do not return duplicate messages, I believe. We always look for a segment with a specific start offset, so the fetch would either return messages from [0, 200] or [101, 200] in your example, depending on what the start offset is. That's just my understanding but could be good to validate that this works as expected.

Reply:
I guess it probably depends on the logic in the caller. The caller may want to enumerate all messages in all segments returned by `Log.logSegments`, which generates duplicate message in this case.

Reply:
Right, I think `Log#read()` should be fine specifically because it returns messages from the first non-empty segment it sees, starting from the given start offset. I made a quick pass-through of other callers and it seems they should be fine with seeing duplicate entries but I need to dig deeper to be able to reason about it. In any case, we can do this as part of a separate JIRA.